rasa/core/featurizers/intent_max_history_tracker_featurizer.py

```diff
@@ -0,0 +1,304 @@
+            positive_label_set = set([labels[0] for labels in positive_label_list])
+
+            for labels in positive_label_list:
+                # Extend the singletone `labels` with the `positive_label_set`.
+                # Note that we need to filter out the redundant label `labels[0]`
+                # from `positive_label_set` so as not to add it twice.
+                filtered_label_set = filter(
+                    lambda label: label != labels[0], positive_label_set
+                )
+                labels.extend(filtered_label_set)
+
+        logger.debug(f"Created {len(example_states)} {self.LABEL_NAME} examples.")
+
+        return example_states, example_labels, example_entities
+
+    def _extract_examples(
+        self,
+        tracker: DialogueStateTracker,
+        domain: Domain,
+        omit_unset_slots: bool = False,
+        ignore_action_unlikely_intent: bool = False,
+    ) -> Iterator[Tuple[List[State], List[Text], List[Dict[Text, Any]]]]:
+        """Creates an iterator over training examples from a tracker.
+
+        Args:
+            trackers: The tracker from which to extract training examples.
+            domain: The domain of the training data.
+            omit_unset_slots: If `True` do not include the initial values of slots.
+            ignore_action_unlikely_intent: Whether to remove `action_unlikely_intent`
+                from training states.
+
+        Returns:
+            An iterator over example states, labels, and entity data.
+        """
+        tracker_states = self._create_states(
+            tracker, domain, omit_unset_slots=omit_unset_slots
+        )
+        events = tracker.applied_events()
+
+        if ignore_action_unlikely_intent:
+            tracker_states = self._remove_action_unlikely_intent_from_states(
+                tracker_states
+            )
+            events = self._remove_action_unlikely_intent_from_events(events)
+
+        label_index = 0
+        for event in events:
+
+            if isinstance(event, ActionExecuted):
+                label_index += 1
+
+            elif isinstance(event, UserUttered):
+
+                sliced_states = self.slice_state_history(
+                    tracker_states[:label_index], self.max_history
+                )
+                label = [event.intent_name or event.text]
+                entities = [{}]
+
+                yield sliced_states, label, entities
+
+    @staticmethod
+    def _cleanup_last_user_state_with_action_listen(
+        trackers_as_states: List[List[State]],
+    ) -> List[List[State]]:
+        """Removes the last state in a tracker if the previous action is action_listen.
+
+        States with the previous action equal to action_listen correspond to states with
+        a new user intent. This information is what UnexpecTEDIntentPolicy is trying to
+        predict so it needs to be removed before obtaining a prediction.
```

Comment:
Sam has an eagle eye for these things, so pointing them out here before he sees it in the final PR 😅 
```suggestion
        States with the previous action equal to `action_listen` correspond to states with
        a new user intent. This information is what `UnexpecTEDIntentPolicy` is trying to
        predict so it needs to be removed before obtaining a prediction.
```